lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
-            extraQueryParameters = Utils.constructUnifiedCacheExtraQueryParameter(idTokenObject, extraQueryParameters);
-            this._silentLogin = true;
-            this.acquireTokenSilent([this.clientId], this.authority, this.getUser(), extraQueryParameters)
-                .then((idToken) => {
-                    this._silentLogin = false;
-                    this._logger.info("Unified cache call is successful");
-                    resolve(idToken);
-                }, (error) => {
-                    this._silentLogin = false;
-                    this._logger.error("Error occurred during unified cache ATS");
-                    this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters);
-                });
-        }
-         else {
-            this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters );
-        }
-      });
-  }
+      // Extract ADAL id_token if it exists
+      var idTokenObject;
+      idTokenObject = this.extractADALIdToken();
 
-  private loginPopupHelper( resolve: any , reject: any, scopes: Array<string>, extraQueryParameters?: string) {
-      //TODO why this is needed only for loginpopup
-      if (!scopes) {
-          scopes = [this.clientId];
+      // silent login if ADAL id_token is retrieved successfully - SSO
+      if (idTokenObject && !scopes) {
+        this._logger.info("ADAL's idToken exists. Extracting login information from ADAL's idToken ");
+        extraQueryParameters = Utils.constructUnifiedCacheExtraQueryParameter(idTokenObject, extraQueryParameters);
+
+        this._silentLogin = true;
+        this.acquireTokenSilent([this.clientId], this.authority, this.getUser(), extraQueryParameters)
+        .then((idToken) => {
+          this._silentLogin = false;
+          this._logger.info("Unified cache call is successful");
+          // TODO: Change resolve to return AuthResponse object
+          resolve(idToken);
+        }, (error) => {
+          this._silentLogin = false;
+          this._logger.error("Error occurred during unified cache ATS");
+          this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters);
+        });
       }
-      const scope = scopes.join(" ").toLowerCase();
-      var popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
-      if (!popUpWindow) {
-          return;
+      // else proceed with login
+      else {
+        this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters );
       }
+    });
+  }
+  
+  /**
+   * Helper function to loginPopup
+   *
+   * @hidden
+   * @param resolve
+   * @param reject
+   * @param scopes
+   * @param extraQueryParameters
+   */
+  private loginPopupHelper( resolve: any , reject: any, scopes: Array<string>, extraQueryParameters?: string) {
+    // TODO: why this is needed only for loginpopup
+    if (!scopes) {
+      scopes = [this.clientId];
+    }
+    const scope = scopes.join(" ").toLowerCase();
```

Comment:
todo, this scopes join is duplicated.

create small util clean cleanScopes or concatScopes so it is done in one place

Reply:
+1, working on in future PR